src/xkbcomp/scanner.c

```diff
@@ -193,16 +193,49 @@ _xkbcommon_lex(YYSTYPE *yylval, struct scanner *s)
     return ERROR_TOK;
 }
 
+/* Basic detection of wrong character encoding based on the first bytes */
+static bool
+check_supported_char_encoding(struct scanner *scanner)
+{
+    /* Skip UTF-8 encoded BOM (U+FEFF) */
+    /* See: https://www.unicode.org/faq/utf_bom.html#bom5 */
+    if (scanner_str(scanner, "\xef\xbb\xbf", 3) || scanner->len < 2) {
+        /* Assume UTF-8 encoding or trivial short input */
+        return true;
+    }
+
+    /* Early detection of wrong file encoding, e.g. UTF-16 or UTF-32 */
+    if (scanner->s[0] == '\0' || scanner->s[1] == '\0') {
+        if (scanner->s[0] != '\0')
+            scanner->token_column++;
+        scanner_err(scanner, "unexpected NULL character.");
+        return false;
+    }
+    /* The first character must be ASCII: whitespace, section, comment */
+    if (!is_ascii(scanner->s[0])) {
+        scanner_err(scanner, "unexpected non-ASCII character.");
+        return false;
+    }
+
+    return true;
+}
```

Comment:
I think the three `check_supported_char_encoding` are the same, right? If so let's move them to `utils.c`? Or make them part of the scanner API as `scanner_check_...`?


After a *very* quick peek at the code, I think we *may* be able to even put this into `scanner_init()` and just fail that for unexpected encodings?

Reply:
Good point. They only differ from their comment about the first character. 

`scanner_init` is also used in `matcher_include` in rules, so I do not see including it there, in order to keep separation of concerns.

I moved it in `scanner-utils.h` as `scanner_check_supported_char_encoding` and moved the aforementioned comments before its use.